### ISSUE TITLE
[iCloud][Binding] - Development of iCloud binding

### ISSUE THREAD

**Issue by pgfeller** [openhab/openhab2-addons#2665]
To make the device information available via iCloud more accessible to OH2 users an iCloud binding would be nice. Altough I do not have experience with JAVA and OSGI I´ll give it a try ... but I´m more than happy for any help.

See also the following forum example/discussion:
https://community.openhab.org/t/icloud-device-data-integration-in-openhab/32329

**Comment by pgfeller** [comment]
A prototype/skeleton is available at https://github.com/pgfeller/openhab2-addons. Any help with how to implement the device discovery and subsequent refresh is highly appreciated.

**Comment by pgfeller** [comment]
Device discovery is now able to add inbox items for every device; but it is not yet possible to create items from them. Probably simething mixed up with thing UIDs and factories ...

**Comment by pgfeller** [comment]
I was able to figure out most of the stuff; discovery, parsing & thing handling works in the prototype. But I´ve no idea how far I´m from best practices (e.g. if such code has a chance ever to see the OH2 repo). Thus I would appreciate some feedback prior to going to the work of declaring all the channels.

**Comment by pgfeller** [comment]
I was able to figure out most of the stuff; discovery, parsing & thing handling works in the prototype. But I´ve no idea how far I´m from best practices (e.g. if such code has a chance ever to see the OH2 repo). Thus I would appreciate some feedback prior to going to the work of declaring all the channels.

**Comment by hmerk** [comment]
@pgfeller 
Hi Patrik,
first of all, thanks again for stepping into this.
I imported your project into my IDE, but was not able to get it running.
Here are my first comments:
Please remove version constraints from MANIFEST.MF
After doing this in my IDE, I get a NPE in updateBridgeChannels, coming from 
content.getPrsId() -> I solved it by changing prsId from int to String in ServerContext.
After that, I got my two devices in inbox, but they won't get any values.

**Comment by pgfeller** [comment]
@hmerk 
Hi Hans,
thank you for your feedback. I´ve commited the changes.
The device data (battery channels only yet) should update on the 1st refresh (default after 5 minutes). Due to timings of the device initialization the data from the discovery is not used - might be possible to improve that - e.g. to update the binding once it registers with the bridge ... if the normal update works I´ll do that.

Do you know if there´s a possibility to declare a "location" channel? It looks like I´ve to provide seperate channels for latitude, longitude - but I´m not sure.

with kind regards,
Patrik

**Comment by hmerk** [comment]
@pgfeller 
Thanks Patrik, it is working after your changes and I already noticed, that the channels will be updated during the first refresh cycle. So far so good.
It should be possible to call the update method once a deviceHandler has registered and fully initialized. But we can have a closer look later on.

A Location channel should be possible, as you declare a channel-type to have an item-type and there is an existing Location item
https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/LocationItem.java

Best
Hans-Jörg

**Comment by pgfeller** [comment]
@hmerk 

To make sure that discovered things have data without the need to wait for the 1st refresh (which will take, depending on configuration) several minutes I use the following code:
```java
 public void childHandlerInitialized(@NonNull ThingHandler childHandler, @NonNull Thing childThing) {
        super.childHandlerInitialized(childHandler, childThing);
        ((iCloudDeviceHandler) childHandler).update(iCloudData.getContent());
    }
```
Unfortunately the data in the inbox is not always updated. I also tried to update the data on channel link; same result … only after the 1st refresh data is shown. Feels like a timing issue/lost event. Any advice?

**Comment by hmerk** [comment]
@pgfeller 
I saw your post in the community, but do not hav a final answer/solution for you.
Will e play with your code later on to see if I will find the right hint for you.

**Comment by pgfeller** [comment]
@hmerk 
Hi Hans,
can you export the binding to a .jar? I would like to give it a try on my real system - but I always get an error when I try to export (cycle reference something ...).

with kind regards,
Patrik

**Comment by pgfeller** [comment]
Was able to export to .jar by closing the runtime projects :-) ...

**Comment by hmerk** [comment]
@pgfeller 
After changing iCloudDeviceHandler initialization to

    @Override
    public void initialize() {
        logger.debug("Initializing iCloud device handler.");
        initializeThing((getBridge() == null) ? null : getBridge().getStatus());
    }

    private void initializeThing(ThingStatus bridgeStatus) {
        logger.debug("initializeThing thing {} bridge status {}", getThing().getUID(), bridgeStatus);

        bridge = (iCloudBridgeHandler) getBridge().getHandler();

        if (bridge != null) {
            if (bridgeStatus == ThingStatus.ONLINE) {
                bridge.registerDevice(this);
                updateStatus(ThingStatus.ONLINE);
            } else {
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
            }
        } else {
            updateStatus(ThingStatus.OFFLINE);
        }

    }

I immediately got battery state after adding devices from inbox. Just lcation was missing.

What I would like to see is a channel with the addresstranslation, like we have here

// Function to transform location coordinates to address
val Functions$Function1<GenericItem, String> locationAddress= [ Coordinates |
	val geocodeURL = "https://maps.googleapis.com/maps/api/geocode/json?latlng=" + iPhone7_Coordinates.state.toString + "&language=german&sensor=true"
	val String geocodeJson = sendHttpGetRequest(geocodeURL)
	var String formattedAddress = transform("JSONPATH", "$.results[0].formatted_address", geocodeJson)
	formattedAddress = formattedAddress.replace(", Germany", "")
	return formattedAddress

Do you think you can add this ???

**Comment by pgfeller** [comment]
@hmerk 

Thank you for the help with the initialization; this indeed did the trick :-).
I´ll include the address lockup next.

with kind regards,
Patrik

**Comment by hmerk** [comment]
@pgfeller 
Great !! Thanks !!

**Comment by pgfeller** [comment]
@hmerk 
Address lookup added to prototype :-)

**Comment by pgfeller** [comment]
@kaikreuzer Thank you for the merge :-); shall I close this issue now?

**Comment by kaikreuzer** [comment]
Sure!

**Comment by kduytsch** [comment]
Hi, not sure if this is the right place, but I have been playing a lot with this bundle, and it was working great. I say "was" because this WE, I changed my timezone (in order to get the right time), and since, the binding doesn't work anymore. 

I changed the /etc/timezone to be Europe/Brussels

The error in openHab.log is 👍 2018-03-13 08:57:17.295 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.thingUpdated()' on 'org.openhab.binding.icloud.handler.ICloudDeviceHandler@12bfcbe': 2018-03-12T18:28:24CET is not in a valid format.
java.lang.IllegalArgumentException: 2018-03-12T18:28:24CET is not in a valid format.

Any idea to workaround this ?
